java-squid/src/main/java/org/sonar/java/resolve/Resolve.java

```diff
@@ -552,26 +552,43 @@ private boolean isArgumentsAcceptable(List<JavaType> argTypes, List<JavaType> fo
   }
 
   private boolean isAcceptableType(JavaType arg, JavaType formal, boolean autoboxing, boolean isParameterized) {
-    // FIXME SONARJAVA-1514 Inference of types for method using parameter types is not handled
-    if (!isParameterized && usesWildcard(formal)) {
+    // FIXME SONARJAVA-1298 type parameters should be handled in formal arguments prior to the call to this method
+    if (!isParameterized && (isParametrizedType(arg) || isParametrizedType(formal) || isWilcardType(arg) || isWilcardType(formal))
+      && !(useTypeVariable(arg) || useTypeVariable(formal))) {
       return types.isSubtype(arg, formal);
     }
     // fall back to behavior based on erasure
     return types.isSubtype(arg.erasure(), formal.erasure()) || (autoboxing && isAcceptableByAutoboxing(arg, formal.erasure()));
   }
 
-  private static boolean usesWildcard(JavaType type) {
-    if (type instanceof JavaType.ParametrizedTypeJavaType) {
-      JavaType.ParametrizedTypeJavaType parametrizedType = (JavaType.ParametrizedTypeJavaType) type;
-      for (JavaType substitution : parametrizedType.typeSubstitution.substitutedTypes()) {
-        if (substitution.isTagged(JavaType.WILDCARD)) {
+  // FIXME SONARJAVA-1298 to be removed
+  private static boolean useTypeVariable(JavaType type) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "useTypeVariable" is 11 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)